File: TPs/TP1/Is/Is.java
/**
 * @author Natanael Barbosa
 */
import java.util.Scanner;

public class Is {
	private static boolean compare(String str1, String str2) {
		boolean res = false;

		if (str1.length() == str2.length()) {
			res = true;
			for (int i = 0; i < str1.length(); i++) {
				res = res && str1.charAt(i) == str2.charAt(i);
			}
		}
		return res;
	}

	/**
	 * Recebe duas strings e retorna quantos caracteres da segunda string contem na primeira. Tira-se as repeticoes dos caracteres na segunda string para nao contar o mesmo caractere mais de uma vez.
	 */
	private static int containsAnyChar(String str1, String str2) {
		int res = 0;
		
		if (str1 != null && str2 != null) {
			String cleaned_str = clean(str2);	
			for (int i = 0; i < cleaned_str.length(); i++) {			
				for (int j = 0; j < str1.length(); j++) {
					if (str1.charAt(j) == cleaned_str.charAt(i)) res++;
					if (res == str1.length()) { j = str1.length(); i = cleaned_str.length();}
				}
			}	
		}
		return res;
	}

	/**
	 * Recebe uma string e retorna outra string sem caracteres repetidos
	 */
	private static String clean(String str) {
		String res = "";
		boolean repeated = false;

		if (str != null) {
			int len = str.length();
			for (int i = 0; i < len; i++) {
				for (int j = i+1; j < len; j++) {
					if (str.charAt(i) == str.charAt(j)) { 
						repeated = true; j=len;
					}
				}
				if (!repeated) { 
					res+=str.charAt(i); 
				}
				
				repeated = false;
			}
		}
		return res;
	}

	/**
	 * Recebe uma string e verifica se todos os seus caracteres sao vogais, retornando true para apenas vogais e false senao.
	 */
	private static boolean isOnlyVowels(String str) {
		boolean res = false;
		int qtyVowels = 0;
		String vowels = "aeiouAEIOU";

		qtyVowels = containsAnyChar(str, vowels);
		if (qtyVowels == str.length() && str.length() != 0) res = true; 

		return res;
	}

	/**
	 * Igual a isOnlyVowels, mas para consoantes
	 */
	private static boolean isOnlyConsonants(String str) {
		boolean res = false;
		int qtyConsonants = 0;
		String consonants = "bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ";

		qtyConsonants = containsAnyChar(str, consonants);
		if (qtyConsonants == str.length() && str.length() != 0) res = true; 

		return res;

	}
	
	/**
	 * Igual a isOnlyVowels, mas para inteiros
	 */
	private static boolean isInteger(String str) {
		boolean res = false;
		int qtdInts = 0;
		String ints = "1234567890";

		qtdInts = containsAnyChar(str, ints);
		if (qtdInts == str.length() && str.length() != 0) res = true; 

		return res;
	}

	/**
	 * Igual a isInteger, mas permitindo um '.' ou ','
	 */
	private static boolean isReal(String str) {
		boolean res = false;
		int qtdNums = 0;
		int dots = 0;
		String nums = "1234567890.,";
 
		qtdNums = containsAnyChar(str, nums);

		if (qtdNums == str.length() && str.length() != 0 && res == false) {
			res = true;
			for (int i = 0; i < str.length(); i++) {
				if (str.charAt(i) == '.' || str.charAt(i) == ',') dots++;
				if (dots > 1) { i = str.length(); res = false; }
			}
		}
		return res;
	}

	public static void main (String[] args) {
		Scanner inp = new Scanner(System.in);
		
		String str = inp.nextLine();

		while (!compare(str, "FIM")) {
			if (isOnlyVowels(str)) System.out.print("SIM "); else System.out.print("NAO ");	
			if (isOnlyConsonants(str)) System.out.print("SIM "); else System.out.print("NAO ");
			if (isInteger(str)) System.out.print("SIM "); else System.out.print("NAO ");
			if (isReal(str)) System.out.println("SIM"); else System.out.println("NAO");
	
			str = inp.nextLine();
		}
		inp.close();
	}
}

File: TPs/TP1/Is_Recursivo/is.c
/**
 * @author Natanael Barbosa
 */
#include <stdio.h>

void get_len(char * str, int * i) {
	if (str[*i] != '\0') {
		(*i)++;
		return get_len(str, i);
	}
}

int compare (char * str1, char * str2) {
	int res = 0;
	int i = 0;
	int j = 0;

	if (str1 != NULL && str2 != NULL) {
		get_len(str1, &i);
		get_len(str2, &j);
		if (i != j) res++;
		else {
			for (int k = 0; k < i; k++) {
				if (str1[k] != str2[k]) { res++; k = i; }
			}
		}
	}
	return res;
}

/**
 * Recebe uma string e conta quantas ocorrencias do caractere especificado acontecem na string.
 */
int count_char (char * str, char c, int i) {
	int res = 0;
	
	if (str[i] != '\0') {
		if (str[i] == c) res++;
		return res + count_char(str, c, i+1);
	}

	return res;
}

/**
 * Recebe uma string e retorna a quantidade de vogais que ela tem.
 */
int is_only_vowels(char * str, int i) {
	int res = 0;
	int n_vowels = 0;
	char * vowels = "aeiouAEIOU";
	int len = 0;

	if (vowels[i] != '\0') {
		get_len(str, &len);
		n_vowels = count_char(str, vowels[i], 0);

		if (n_vowels < len) return n_vowels + is_only_vowels(str, i+1);
		else return n_vowels;
	}
}

/**
 * O mesmo que is_only_vowels, mas para consoantes.
 */
int is_only_consonants(char * str, int i) {
	int res = 0;
	int n_consonants = 0;
	char * consonants = "bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ";
	int len = 0;

	if (consonants[i] != '\0') {
		get_len(str, &len);
		n_consonants = count_char(str, consonants[i], 0);

		if (n_consonants < len) return n_consonants + is_only_consonants(str, i+1);
		else return n_consonants;
	}
}

/**
 * O mesmo que is_only_vowels, mas para numeros.
 */
int is_int(char * str, int i) {
	int res = 0;
	int n_ints = 0;
	char * ints = "1234567890";
	int len = 0;

	if (ints[i] != '\0') {
		get_len(str, &len);
		n_ints = count_char(str, ints[i], 0);

		if (n_ints < len) return n_ints + is_int(str, i+1);
		else return n_ints;
	}
}

/**
 * O mesmo que is_int, mas permitindo ',' e '.'
 */
int is_real(char * str, int i) {
	int res = 0;
	int n_nums = 0;
	char * nums = "1234567890.,";
	int len = 0;

	if (nums[i] != '\0') {
		get_len(str, &len);
		n_nums = count_char(str, nums[i], 0);

		if (n_nums < len) return n_nums + is_real(str, i+1);
		else return n_nums;
	}
}

/**
 * Retorna 1 caso a string tenha apenas um '.' ou ',', e 0 senao
 */
int count_dots(char * str) {
	int res = 1;
	int count = 0;
	
	count = count_char(str, '.', 0);
	if (count <= 1) count += count_char(str, ',', 0);
	if (count > 1) res--;

	return res;
}


int main () {
	char str [1000];
	int len = 0;

	scanf("%[^\n]", str); getchar();
	while (compare(str, "FIM") != 0) {
		get_len(str, &len);

		if (is_only_vowels(str, 0) == len) printf("SIM ");
		else printf("NAO ");
		
		if (is_only_consonants(str, 0) == len) printf("SIM ");
		else printf("NAO ");

		if (is_int(str, 0) == len) printf("SIM ");
		else printf("NAO ");	

		if (is_real(str, 0) == len && count_dots(str) == 1) printf("SIM\n"); // aqui, o valor para ser real deve ter um ',' ou '.' e apenas numeros
		else printf("NAO\n");
		
		scanf("%[^\n]", str); getchar();
		len = 0;
	}

	return 0;
}
